File: Strivers-A2Z-DSA-Sheet/06.Bit Manipulation/3. Advanced Maths/javascript/02. All divisors of number.js
/**
QUESTION:
Given an integer N, return all its divisors in ascending order.

Approach:
Iterate i from 1 to sqrt(N). For each divisor i, push it immediately. If N / i differs from i, store it for later and finally append the stored values in reverse to maintain ascending order.

Time Complexity: O(sqrt(N))
Space Complexity: O(sqrt(N)) for temporary storage
*/

function allDivisors(n) {
    const high = [];
    const result = [];
    for (let i = 1; i * i <= n; i++) {
        if (n % i === 0) {
            result.push(i);
            if (n / i !== i) {
                high.push(n / i);
            }
        }
    }
    for (let j = high.length - 1; j >= 0; j--) {
        result.push(high[j]);
    }
    return result;
}

module.exports = { allDivisors };

File: Strivers-A2Z-DSA-Sheet/06.Bit Manipulation/3. Advanced Maths/python/02. All divisors of number.py
"""QUESTION:
Given an integer n, return all of its divisors in ascending order.

Approach:
Iterate i from 1 to floor(sqrt(n)). Append i immediately when it divides n. If n / i differs from i, store it for later. Once the loop finishes, append the stored values in reverse order so the result stays sorted.

Time Complexity: O(sqrt(n))
Space Complexity: O(sqrt(n)) for temporary storage
"""
from typing import List

def all_divisors(n: int) -> List[int]:
    result: List[int] = []
    high: List[int] = []
    i = 1
    while i * i <= n:
        if n % i == 0:
            result.append(i)
            if n // i != i:
                high.append(n // i)
        i += 1
    result.extend(reversed(high))
    return result
